File: lib/models/filter_settings.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'filter_settings.freezed.dart';
part 'filter_settings.g.dart';

@freezed
abstract class FilterSettings with _$FilterSettings {
  const FilterSettings._();

  const factory FilterSettings({
    @Default(true) bool onlyVideo,
    @Default(false) bool ignoreSmallFiles,
    int? minSizeBytes,
    int? downloadLimit,
    @Default(<String>{'mp4', 'webm'}) Set<String> allowedExtensions,
  }) = _FilterSettings;

  factory FilterSettings.fromJson(Map<String, dynamic> json) =>
      _$FilterSettingsFromJson(json);

  static const int defaultSmallFileThresholdBytes = 2 * 1024 * 1024;

  int? get effectiveMinSizeBytes {
    if (minSizeBytes != null) return minSizeBytes;
    return ignoreSmallFiles ? defaultSmallFileThresholdBytes : null;
  }

  bool acceptsExtension(String extension) {
    final normalized = extension.toLowerCase().replaceFirst('.', '');
    if (onlyVideo && !{'mp4', 'webm'}.contains(normalized)) return false;
    return allowedExtensions.contains(normalized);
  }

  bool acceptsSize(int? sizeBytes) {
    final minimum = effectiveMinSizeBytes;
    if (minimum == null || sizeBytes == null) return true;
    return sizeBytes >= minimum;
  }
}

File: app/src/main/java/io/github/rustnomicon/sourceinstaller/data/model/FilterSettings.kt
package io.github.rustnomicon.sourceinstaller.data.model

/**
 * User-adjustable media filters. Port of the Flutter `FilterSettings` model.
 */
data class FilterSettings(
    val onlyVideo: Boolean = true,
    val ignoreSmallFiles: Boolean = false,
    val minSizeBytes: Long? = null,
    val downloadLimit: Int? = null,
    val allowedExtensions: Set<String> = setOf("mp4", "webm"),
) {
    val effectiveMinSizeBytes: Long?
        get() = minSizeBytes
            ?: (if (ignoreSmallFiles) DEFAULT_SMALL_FILE_THRESHOLD_BYTES else null)

    fun acceptsExtension(extension: String): Boolean {
        val normalized = extension.lowercase().removePrefix(".")
        if (onlyVideo && normalized !in VIDEO_EXTENSIONS) return false
        return normalized in allowedExtensions
    }

    fun acceptsSize(sizeBytes: Long?): Boolean {
        val minimum = effectiveMinSizeBytes ?: return true
        if (sizeBytes == null) return true
        return sizeBytes >= minimum
    }

    companion object {
        const val DEFAULT_SMALL_FILE_THRESHOLD_BYTES: Long = 2L * 1024 * 1024
        val VIDEO_EXTENSIONS = setOf("mp4", "webm")
        val ALL_EXTENSIONS = setOf("mp4", "webm", "jpg", "jpeg", "png", "gif", "webp")
    }
}
